### ISSUE TITLE
Table system update of keywordset does not respect lock and set table.dat to 0 size

### ISSUE THREAD

**Issue by kgolap** [casacore/casacore#1077]
When multiple processes try to update keywords in a disk based table  the lock on the on the table is not respected.
So the keywords update suffers from race conditions ..sometimes part gets updated or not.
Also, more importantly,  it has been noted that when updating the keywords the other processes that are trying to access the table would throw an exception that table.dat size is now 0. @cquike traced it to the file being open with O_TRUNC somehow.
This appears for example when one process is trying to do PagedImage::setimageinfo and others are trying to read part of the data.

**Comment by cquike** [comment]
Note that table.dat is open with O_TRUNC and then filled again with the proper information. So the problem of the race condition happens when a process just opens with O_TRUNC and the second process tries to access table.dat *before* the first process had the chance to properly fill table.dat.

**Comment by cquike** [comment]
I can think of two possible solutions:

1. The constructor of Table checks for the existence of a lock *before* trying to open table.dat. Right now it tries to open table.dat to verify that this is actually a table, by reading the first bytes of the file and looking for some "magic number". This test could easily be done after grabbing the lock. One can even say that the existence of the lock is in itself enough to guarantee that we are dealing with a table.
2. When updating the keywords, rather than opening with O_TRUNC one could open normally and seek the file descriptor to the place where the keywords are written. That should leave the first bytes where the magic number is intact and the check performed by the second process will succeed.

**Comment by cquike** [comment]
I attach a code snippet that nicely shows the problem (github doesn't allow me to upload the file with .cc extension, therefore the .txt, sigh....). You have to compile with mpi support. To run it:
$ mkdir MYTMP
$ /opt/casa/02/bin/mpirun -np 4 tPagedImage
I am rank 2 of 4
I am rank 3 of 4
I am rank 0 of 4
I am rank 1 of 4
Exception caught: FilebufIO::readBlock - incorrect number of bytes read for file /scratch/cgarcia/build/casa-el7-ts2-dev02/casa/casacore/MYTMP/tPagedImage_tmp.img1/table.dat

Basically the program updates the image info in process0 and the other ones try to open the image. If you increase the time each secondary process waits (line 78) the problems disappears (in my case with 300000 nsec), which demonstrates that this is a race condition.


[tPagedImage.cc.txt](https://github.com/casacore/casacore/files/5667820/tPagedImage.cc.txt)

**Comment by cquike** [comment]
Hi,
kindly pinging whether someone as look into this problem. @tammojan , @gervandiepen : this is hitting CASA in different ways, and although there have been workarounds put in place in the worst cases we still see this problem under our testing.

**Comment by cquike** [comment]
I believe this can be closed after #1078 was merged.
